### ISSUE TITLE
Objective-C linker is used when project contains C++ sources

### ISSUE THREAD

**Issue by oleavr** [mesonbuild/meson#1766]
Which results in linker errors. It seems like the correct thing to do in
such cases would be to use the `objcpp` linker.

**Comment by jpakkane** [comment]
As an extra bonus, the user might not have requested ObjC++ as a language. Which kind of makes sense if they are not using it directly.

🤔

**Comment by dcbaker** [comment]
@nirbeek and I ran into this for C and C++ not long ago. The solution that was implemented was to order the C-like languages so that it would "promote" as necessary. Currently the order (excluding irrelevant to this discussion languages) is [objcpp, objc, cpp, c]. Since (I think) GCC and Clang are the only compilers that implement ObjC and ObjC++ it might be sufficient to make cpp supersede objc in that list.

I did add a nice error message about missing languages, so there's that.

**Comment by Volker-Weissmann** [comment]
This was fixed by 6f3e2a0a0753ad0a4fc84a74fdc6209f436d4371
Somebody please close this issue

Test on current master (f71c9aebfbceb8623756b981392b6d6d2b4e7c22):

```meson
project('hello', ['objc', 'cpp'])
executable('hello', ['empty.m', 'hello.cpp'])
```

```
[1/3] cc -Ihello.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -MD -MQ hello.p/empty.m.o -MF hello.p/empty.m.o.d -o hello.p/empty.m.o -c ../empty.m
[2/3] c++ -Ihello.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -MD -MQ hello.p/hello.cpp.o -MF hello.p/hello.cpp.o.d -o hello.p/hello.cpp.o -c ../hello.cpp
[3/3] c++  -o hello hello.p/empty.m.o hello.p/hello.cpp.o -Wl,--as-needed -Wl,--no-undefined
```

**Comment by eli-schwartz** [comment]
Yeah, this was re-reported as #2468 and fixed there...
